KIPs/kip-114.md

```diff
@@ -0,0 +1,324 @@
+def get_proposer_sk(num):
+    proposerIdx = num % validatorNum
+    return proposerBLSPrivKeys[proposerIdx]
+
+
+def gen_next_header(parent) -> Header:
+    nextBlockNum = parent.number + 1
+    if is_xxxx_fork_enabled(nextBlockNum):
+        sk = get_proposer_sk(nextBlockNum)
+
+        randomReveal = calc_random_reveal(sk, nextBlockNum)
+
+        if nextBlockNum == FORK_BLOCK:
+            # parent mixHash does not exist, so fill with default
+            parentMixHash = b"\x00" * 32
+        else:
+            parentMixHash = parent.mixHash
+
+        mixHash = calc_mix_hash(parentMixHash, randomReveal)
+    else:
+        randomReveal = b""
+        mixHash = b""
+
+    return Header(nextBlockNum, randomReveal, mixHash)
+
+
+def keccak256(msg):
+    keccak256 = keccak.new(digest_bits=256)
+    keccak256.update(msg)
+    return keccak256.digest()
+
+
+def block_num_to_bytes(num) -> bytes:
+    return num.to_bytes(8, byteorder="big")
+
+
+def calc_random_reveal(sk, num) -> bytes:
+    msg = block_num_to_bytes(num)
+    return bytes(PopSchemeMPL.sign(sk, msg))
+
+
+def calc_mix_hash(parentMixHash, randomReveal):
+    return xor(parentMixHash, keccak256(randomReveal))
+
+
+def xor(a: bytes, b: bytes):
+    return bytes(x ^ y for x, y in zip(a, b))
+
+
+def verify_xxxx_header(header, parentMixHash):
+    if not is_xxxx_fork_enabled(header.number):
+        return True
+
+    # pop verify
+    blsPublicKeyInfo = BLSRegistry().getInfo(header.proposer)
+    publicKey = G1Element.from_bytes(blsPublicKeyInfo.publicKey)
+    pop = G2Element.from_bytes(blsPublicKeyInfo.pop)
+    if not PopSchemeMPL.pop_verify(publicKey, pop):
+        return False
+
+    # sig verify
+    msg = block_num_to_bytes(header.number)
+    sig = G2Element.from_bytes(header.randomReveal)
+    if not PopSchemeMPL.verify(publicKey, msg, sig):
+        return False
+
+    # mixHash verify
+    if header.number == FORK_BLOCK:
+        # parent mixHash does not exist, so fill with default
+        parentMixHash = b"\x00" * 32
+    mixHash = calc_mix_hash(parentMixHash, header.randomReveal)
+    return mixHash == header.mixHash
+
+
+def main():
+    header = Header(FORK_BLOCK - 2, b"\x04" * 96, b"\x00" * 32)
+    prevHeader = header
+    N = 10
+    # print header numbers in [FORK_BLOCK - 1, FORK_BLOCK + N]
+    for _ in range(N + 2):
+        header = gen_next_header(header)
+        print(header)
+        verified = verify_xxxx_header(header, prevHeader.mixHash)
+        assert verified
+        print("verified:", verified)
+        print("=" * 80)
+        prevHeader = header
+
+
+main()
+```
+
+### Genesis Block
+
+If `FORK_BLOCK` is zero, the genesis block must have `randomReveal`, `mixHash` fields.
+
+## Rationale
+
+### Names of the header fields
+
+The naming is highly inspired by [EIP-4339](https://eips.ethereum.org/EIPS/eip-4399).
+We use `randomReveal` instead of RANDAO reveal because the randoms are generated in a different manner.
```

Comment:
What is the field name of Ethereum?

Reply:
In [spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#beaconblockbody), it's `randao_reveal`.
In [EIP](https://eips.ethereum.org/EIPS/eip-4399), the term `RANDAO reveal` is used.

Reply:
Is `randao_reveal` is included in the block header? in consensus layer block, or execution layer block?

Reply:
It's in consensus layer's block